src/buffer4/worker.rs

```diff
@@ -225,3 +226,11 @@ impl Handle {
             .unwrap_or_else(|| Closed::new().into())
     }
 }
+
+async fn option_and_then<T, U, F, Fut>(x: Option<T>, f: F) -> Option<U>
+where
+    F: FnOnce(T) -> Fut,
+    Fut: Future<Output = Option<U>>,
+{
+    f(x?).await
+}
```

Comment:
this is specifically used for calling `recv`, so I think it isn't _really_ necessary to take closure. it could just be:
```suggestion
async fn recv_option<T>(x: Option<&mut mpsc::Receiver<Option<T>>>)) -> Option<T> {
    x?.recv().await
}
```